doc/limit/SONiC Limit user session.md

```diff
@@ -0,0 +1,222 @@
+
+# 2 Configuration and Management Requirements
+## 2.1 SONiC CLI
+ - Manage login session or memory  limit settings
+```
+    config limit login { enable | disable }
+    config limit login { add | del } {user | group | global} <name> <number>
+```
+ - Show limit
+```
+    show limit login
+```
+
+## 2.2 Config DB
+ - Login limit are fully configurable by config DB.
+
+# 3 Design
+ - Design diagram:
+
+```mermaid
+graph LR;
+%% SONiC CLI update config DB
+CLI[SONiC CLI] -- update limit setting --> CONFDB[(Config DB)];
+
+
+%% HostCfgd subscribe config DB change
+CONFDB -.-> HOSTCFGD[Hostcfgd];
+
+%% HostCfgd Update config files
+HOSTCFGD -- update limits.conf --> PAMCFG[limits.conf];
+
+%% pam_limits.so will handle login limit
+PAMCFG -.-> LIMITLIB[pam_limits.so];
+LIMITLIB -- login --- USERSESSION(user session);
+
+```
+
+## 3.1 Login limit Implementation
+ - Enable PAM plugin pam_limits.so to support login limit.
+ - When login limit exceed, pam_limits.so will terminate login session with error message.
+
+```mermaid
+sequenceDiagram  
+
+%% user login without exceed the limit
+ SSH/Console->>SSHD  : login
+	 activate  SSHD
+	 SSHD->>pam_limits.so: check login limit
+		 activate  pam_limits.so
+		 pam_limits.so->>bash: not exceed the limit
+			 activate  bash
+			 bash->>SSH/Console: login success
+			 deactivate  bash
+	 SSH/Console->>SSHD  : logout
+	 deactivate  SSHD
+ 
+%% user login exceed the limit
+ SSH/Console->>SSHD  : login
+	 activate  SSHD
+	 SSHD->>pam_limits.so: check login limit
+		 activate  pam_limits.so
+		 pam_limits.so->>SSH/Console: exceed the limit
+		 deactivate  pam_limits.so
+	 deactivate  SSHD
+```
+ #### Other solution for Linux login session limit
+
+|                   | How                                                | Pros                                                         | Cons                       |
+| ----------------- | -------------------------------------------------- | ------------------------------------------------------------ | -------------------------- |
+| PAM limit         | Change PAM setting file: /etc/security/limits.conf | Support per-user/per-group/global limit. Only need change config file. |                            |
+| Bash login script | Call script when user login                        |                                                              | Need develop new script.   |
+| SSHD config       | Change SSHD setting file: /etc/sshd_config         |                                                              | Only support global limit. |
+
+- SONiC will create new user when domain user login, PAM limit support config limit to a not existed user.
+
+
+
+## 3.2 Default login session limitation Implementation
+- Global max login sessions: 10
```

Comment:
>10 [](http://example.com/codeflow?start=29&length=2)

I am worried about this tight limitation. If a high sev incident happens, there would be much more login users on the device. What happens if more than 10 login sessions? Prevent new login or kickoff old login?

Will login itself be a problem? Will a login session occupy a lot of memory?

Reply:
When login more than 10 session, then user can't create new session, and will get login session been limited message when login. old session will not be kickoff.

Login itself will be a problem: every login session will take 30 MB memory, so 10 login session will take 300MB memory.

Reply:
Hua: check if we can limit login by current free memory.

Reply:
I checked the pam_limit, it not support limit login by current free memory. however we can create new pam plugin or patch it for the memory limit.
global limit removed.